Improve thermistor 20-21 descriptions re: ADC Vref and INA826 PT100 supply V

Quas7 opened the pull request:
**Description**
I removed the 5V to 3.3V table #define OV_SCALE conversion. In essence, thermistor_21.h has now the same code content as thermistor_20.h.

The conversion is leading to an erronous temperature reading, as e.g. a 1°C reading at room temperature (see issues #16561 and #19234)

The table provides temperature reference points in terms of ADC-counts. As long as the ADC reference voltage (LPC, STM normally 3.3V) and the INA826 supply voltage (here as well 3.3V) are identical, there is no theoretical difference in ADC counts between 5V and 3V3 setups (besides noise, etc.).

Little more background:
the ADC is providing counts in relation to his reference voltage and the INA826 opamp does the very same in the analog domain referring to its supply voltage. Therefore, same references on both ends cancel out for the look-up table.

I validated with an additional PT100 measurement for 100°C and room temperature.

Note, that the quantization error of the table in combination with rounding of the temperature reading might lead to a few degrees (seen up to 3°C) of error around the center between datapoints, e.g. 25°C, 35°C, etc.

**Benefits**
Temp Probe Nr. 21 gets usable for INA826 PT100 in 3V3 environments. 
Currently Nr. 21 is greatly underestimating the hotend temperature.

**Configurations**
in any Configuration.h with INA826-PT100 and 3V3 board setup:
#define TEMP_SENSOR_0 21

(I cannot attach .h files here?!)
[Configuration.zip](https://github.com/MarlinFirmware/Marlin/files/5168291/Configuration.zip)


**Related Issues**
#19234
#16561

Quas7 committed:
Update thermistor_21.h

Removed 5V to 3.3V table conversion.
The conversion is leading to an erronous temperature reading, as for instance 1°C reading at room temperature.
The 5V table provided in _ADC-counts_ is valid as long as the ADC reference voltage and the INA826 supply voltage are identical.
The ADC is providing counts in relation to his reference voltage and the INA826 does the very same in the analog domain.
Validated with additional PT100 measurement for 100°C and room temperature.
Note, that the quantization error of the table in combination with rounding of the temperature reading might lead to a few degrees (seen up to 3°C) of error around the center between datapoints, e.g. 25°C, 35°C, etc.

Quas7 committed:
Update Configuration.h

added temp_sensor_0 21 for integration test

AnHardt commented on Marlin/Configuration.h:
```diff
@@ -416,7 +416,7 @@
  */
-#define TEMP_SENSOR_0 1
+#define TEMP_SENSOR_0 21
 #define TEMP_SENSOR_1 0
```
Please remove your config change.

Quas7 replied:
done.

AnHardt commented:
Having two times the same table is completely pointless.
21 is ought to be used when the module is powered with 5V but the ADC-reference-voltage is 3.3V. 
Only the descriptions could be better - less misunderstandeble.

Quas7 commented:
@AnHardt that makes absolutely sense, as the conversion is doing exactly this. 
I would also opt for adding a remark for No. 21 as using 5V on a 3V3 ADC is not risk free (not 5V tolerant at least for LPC).

Correcting the description is then the correct way and adding a comment on No. 20, that it can be used for "symmetric" setups in terms of supply and references.

Quas7 committed:
Update Configuration.h

changed back for PR

Quas7 commented:
20 : Pt100 with circuit in the Ultimainboard V2.x with mainboard ADC reference voltage = INA826 amplifier-board supply voltage
21 : Pt100 with circuit in the Ultimainboard V2.x with 3.3v ADC reference voltage (STM32, LPC176x....) and 5V INA826 amplifier board-supply for increased temperature resolution (note: ADC pins are not 5V tolerant)

I could prepare some simple documentation how to add a 3.6V Zener diode between INA-out and GND to protect the ADC.

Bob-the-Kuhn commented:
The chances of a 5V INA826 damaging a 3.3V analog input is very low.

The temperature needed to get above a 3.3V output is way beyond normal operation.  There is a parasitic diode from the AIN pad to 3.3V but the INA826 has so little muscle that even at full output it won't be able to drive enough current through it to cause any thermal damage.

I've posted the details on this in a previous thread.  Unfortunately a quick search failed to turn it up.

---

A couple of thoughts on the 20 & 21 notes.

I'd drop the "for increased temperature resolution" part.  I seriously doubt that this is of any real practical value.  I've switched to a PT100 without the amplifier which has less resolution.  I haven't noticed  any difference in the "noise" in the thermal history graphs of PT100 with amplifier, PT100 without amplifier and 100K NTC thermistors.

My recomendation for the notse would be:
20 : Pt100 with circuit in the Ultimainboard V2.x with mainboard ADC reference voltage = INA826 amplifier-board supply voltage (notes: 1) Must use an ADC input without a pull up resistor. 2) Some INA826 amplifiers do not operate properly at 3.3V.  In this case consider using thermal tables 147 [PT100, 4.7K pullup, no amplifier], 110 [PT100, 1K pullup, no amplifier], or 21 [PT100, no pullup, 5V powered amplifier])
21 : Pt100 with circuit in the Ultimainboard V2.x with 3.3v ADC reference voltage (STM32, LPC176x....) and 5V INA826 amplifier board supply. (note: ADC pins are not 5V tolerant.  Not recomended as there is a small chance of damaging the CPU.)

AnHardt commented:
Boss text version is already very good. Maybe it could be improved by adding the 'border' temperature where 21 becomes dangerous for the 3.3V ADC. Something like: "Don't use this above ???°C !", where ??? could be 500 - that would leave a bit of leeway. On the other hand almost all wire-isolation of cheap PT100 has given up then.

Quas7 commented:
@Bob-the-Kuhn 
I agree, that during normal operation (<500°C) there is no chance to get beyond 3.6V (LPC max rating).
Also, I checked the INA826 datasheet and you are absolutely right as it only provides 16mA as short circuit current. 
I did not find the reference to the Vdd-up diode but I assume it is a standard bulk ESD diode shown in some AppNote for STM32 or LPC. Worst case at 5V INA-out would result in ~2.7V drop over the diode and 43mW of power dissipation, which is of no concern for bulk diodes. 
Anyway, I added a 3.6V Zener now, just in case my cheap China-clone INA-board goes crazy. ;)

Regarding the resolution and proper scaling for the ADC with 5V vs 3v3 I quickly ran some double testing back and forth with prope no. 21 vs 20 at 3V3 vs. 5V INA-supply.
I got 24.49°C vs. 25.31°C and a seperate PT100 probe reads 23.7°C at the heating block.
Therefore, resolution was increased to 0.2°C compared to 0.31°C or +33%, as expected. And the error was cut in half with 3% (+0.79°C) vs. 6% (1.61°C). This fits my expectation, as we get around +/-6.25% error between the table datapoints as result of the quantization and we move 33% up in the table with 5V INA-supply.

All in all, I agree, that this is an academic (cosmetic) benefit of low practical value and the higher resolution might only be of interest for highly tuned PID controls. 

The text for the propes would be fine for me. 

@AnHardt 
 Adding the upper temperature limit is not really required but you are right as there will always be another smart guy starting an issue or PR for it, if it is missing (like myself in this case). ;P

Quas7 committed:
Update Configuration.h

updated description for temperature probe 20 and 21

Quas7 committed:
Update thermistor_21.h

reverted back to original code and added additional comment on intended usecase for this probe (5V INA826 supply, 3v3 ADC ref on mainboard)

Quas7 commented:
as discussed here https://github.com/MarlinFirmware/Marlin/issues/19234, there might be a masking of a thermal runaway, if overtemp is set beyond possible max ADC value with 5V INA-supply and 3V3 ADC-reference.

In my opinion, multiple values (=hit-counter) close to the full scale readout on the ADC should trigger a overtemp warning. 
This should be set for all temp probes as fire prevention, if the user misconfigures overtemp.
Maybe there is already something like this implemented (did not test)?

thinkyhead committed:
Update Configuration.h

thinkyhead committed:
Prettify

thinkyhead committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into pr/19246
